src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+                    a = i;
+                }
+            } else {
+                em[i] = INFINITY;
+            }
+        }
+
+        let mut b = replace(&mut pi[a], n);
+        let mut c = replace(&mut lambda[a], em[a]);
+
+        if a < n - 1 {
+            while  b < n - 1 {
+                b = replace(&mut pi[b], n);
+                c = replace(&mut lambda[b], c);
+            }
+
+            if b == n - 1 {
+                pi[b] = n;
+                lambda[b] = c;
+            }
+        }
+
+        for i in (0..n) {
+            if pi[pi[i]] == n && lambda[i] >= lambda[pi[i]] {
+                pi[i] = n;
+            }
+        }
+    })
+}
+
+
+fn using_pointer_representation<'a, P, F>(points: &Vec<&'a P>, f: F) -> Box<Dendrogram<&'a P>>
+    where P: Point, F: Fn(usize, &mut Vec<usize>, &mut Vec<f64>, &mut Vec<f64>)
+{
+    let mut pi = vec![0; points.len()];
+    let mut lambda = vec![0.0; points.len()];
+    let mut em = vec![0.0; points.len()];
+
+    pi[0] = 0;
+    lambda[0] = INFINITY;
+
+    for n in (1..points.len()) {
+        pi[n] = n;
+        lambda[n] = INFINITY;
+
+        for i in (0..n) {
+            em[i] = points[i].dist(points[n]);
+        }
+
+        f(n, &mut pi, &mut lambda, &mut em);
+    }
+
+    // convert pointer representation to dendrogram
+    let mut dendrograms: Vec<_> = (0..points.len()).map(|i|
+        Some(Box::new(Dendrogram::Leaf(points[i])))).collect();
+
+    let mut idx: Vec<_> = (0..points.len()).collect();
+    idx.sort_by(|&a, &b| lambda[a].partial_cmp(&lambda[b]).unwrap());
+
+    for i in (0..points.len()-1) {
+        let leaf = idx[i];
+        let merge = pi[leaf];
+
+        let a = replace(&mut dendrograms[leaf], None).unwrap();
+        let b = replace(&mut dendrograms[merge], None).unwrap();
+
+        if leaf < merge {
+            dendrograms[merge] = Some(Box::new(Dendrogram::Branch(lambda[leaf], a, b)));
+        } else {
+            dendrograms[leaf] = Some(Box::new(Dendrogram::Branch(lambda[leaf], b, a)));
+        }
+    }
+
+    dendrograms.swap_remove(points.len() - 1).unwrap()
+}
+
+
+#[cfg(test)]
+#[allow(non_snake_case)]
+fn minimal_distance(d: &Fn(usize, usize) -> f64, A: &[usize], B: &[usize]) -> f64 {
+    A.iter().flat_map(|&a|
+        B.iter().map(move |&b| Dist(d(a,b)))
+    ).min().unwrap().0
+}
+
+#[allow(non_snake_case)]
+fn maximal_distance(d: &Fn(usize, usize) -> f64, A: &[usize], B: &[usize]) -> f64 {
+    A.iter().flat_map(|&a|
+        B.iter().map(move |&b| Dist(d(a,b)))
+    ).max().unwrap().0
+}
+
+#[derive(PartialEq, PartialOrd)]
+struct Dist(f64);
+impl Eq for Dist { }
+impl Ord for Dist {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+
+/// A hierarchical tree structure describing the merge operations.
+#[derive(Clone, Debug)]
+pub enum Dendrogram<T> {
+    /// The union of two sub-trees (clusters) and their distance.
+    Branch(f64, Box<Dendrogram<T>>, Box<Dendrogram<T>>),
```

Comment:
I keep coming back to this, and I'm not super keen on this data representation: the number of `Box`s here is quite large! I wonder if this could be something like:

```rust
pub enum DendrogramElement<T> {
     Branch(f64, usize, usize),
     Leaf(T)
}
```

with the whole dendrogram is represented by a `Vec<DendrogramElement<T>>`. The `usize`s would be indexes into the `Vec`. In a perfect world these representations are pretty much identical, but the `Vec` version has e.g. better locality, and fewer allocations, so less allocator metadata is required, and one can allocate the whole thing up-front (the number of `DendrogramElement<T>`s is pretty much 2n, I believe?), instead of having to repeatedly hit the allocator/find new memory.

That said, this could easily be not actually a problem in practice. I suppose I could do some benchmarks/experimentation/profiling.